pkg/auth/providers/saml/saml_provider.go

```diff
@@ -0,0 +1,287 @@
+package saml
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/apis/management.cattle.io/v3public"
+	"github.com/rancher/types/client/management/v3"
+	publicclient "github.com/rancher/types/client/management/v3public"
+	"github.com/rancher/types/config"
+	"github.com/rancher/types/user"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	Name      = "saml"
+	userType  = Name + "_user"
+	groupType = Name + "_group"
+)
+
+type Provider struct {
+	ctx         context.Context
+	authConfigs v3.AuthConfigInterface
+	userMGR     user.Manager
+	samlClient  *SPClient
+}
+
+var Samlp *Provider
+
+func (s *Provider) GetName() string {
+	return Name
+}
+
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userMGR user.Manager) common.AuthProvider {
+	samlClient := &SPClient{}
+	Samlp = &Provider{
+		ctx:         ctx,
+		authConfigs: mgmtCtx.Management.AuthConfigs(""),
+		userMGR:     userMGR,
+		samlClient:  samlClient,
+	}
+	return Samlp
+}
+
+func (s *Provider) CustomizeSchema(schema *types.Schema) {
+	schema.ActionHandler = s.actionHandler
+	schema.Formatter = s.formatter
+}
+
+func (s *Provider) TransformToAuthProvider(authConfig map[string]interface{}) map[string]interface{} {
+	p := common.TransformToAuthProvider(authConfig)
+	p[publicclient.SamlProviderFieldRedirectURL] = formSamlRedirectURLFromMap(authConfig)
+	return p
+}
+
+func (s *Provider) AuthenticateUser(input interface{}) (v3.Principal, []v3.Principal, map[string]string, error) {
+	login, ok := input.(*v3public.CodeBasedLogin)
+	if !ok {
+		return v3.Principal{}, nil, nil, errors.New("SAML: unexpected input type")
+	}
+
+	return s.LoginUser(login, nil, false)
+}
+
+func (s *Provider) LoginUser(samlCredential *v3public.CodeBasedLogin, config *v3.SamlConfig, test bool) (v3.Principal, []v3.Principal, map[string]string, error) {
```

Comment:
i know you just copied from github provider, but this function doesnt need to be exported

Reply:
In general, can you review all your functions and ensure you are only exporting ones that actually need to be called from a different package?